src/pymoca/tree.py

```diff
@@ -259,6 +273,694 @@ def handle_walk(self, listener: TreeListener, tree: Union[ast.Node, dict, list])
 
 
+def find_name(
+    name: Union[str, ast.ComponentRef],
+    scope: ast.Class,
+    copy: bool = True,
+    search_imports: bool = True,
+    search_parent: bool = True,
+    check_builtin_classes=False,
+    current_extends: Optional[Set[ast.ExtendsClause]] = None,
+) -> Optional[Union[ast.Class, ast.Symbol]]:
+    """Modelica name lookup on a tree of ast.Class and ast.InstanceClass
+
+    Look for ast.Class or ast.Symbol per the Modelica Language Specification Version 3.5:
+    1. Simple Name Lookup (spec 5.3.1)
+        0. Predefined types (`Real`, `Integer`, `Boolean`, `String`) (spec 4.8)
+        1. Iteration variables
+        2. Classes
+        3. Components (Symbols in Pymoca)
+        4. Classes and Components from Extends Clauses
+        5. Qualified Import names, see 4 (but not from Extends Clauses) (spec 13.2.1)
+        6. Public Unqualified Imports (error if multiple are found) (spec 13.2.1)
+        7. Repeat 1-6 for each lexically enclosing instance scope, stopping at `encapsulated`
+        unless predefined type, function, operator. If name matches a variable (a.k.a. component,
+        symbol) in an enclosing class, it must be a `constant`.
+    2. Composite Name Lookup (e.g. `A.B.C`) (spec 5.3.2)
+        1. `A` is looked up using Simple Name Lookup
+        2. If `A` is a Component:
+            1. `B.C` is looked up from named component elements of `A`
+            2. If not found and if `A.B.C` is used as a function call and `A` is a scalar or can be
+            evaluated as a scalar from an array and `B` and `C` are classes,
+            it is a non-operator function call.
+        3. If `A` is a Class:
+            1. `A` is temporarily flattened without modifiers
+            2. `B.C` is looked up among named elements of temp flattened class,
+            but if `A` is not a package, lookup is restricted to `encapsulated` elements only
+            and "the class we look inside shall not be partial in a simulation model".
+    3. Global Name Lookup (e.g. `.A.B.C`) (spec 5.3.3)
+        1. `A` is looked up in global scope. (`A` must be a class or a global constant.)
+        2. If `A` is a class, follow procedure 2.3.
+    4. Imported Name Lookup (e.g. `A.B.C`, `D = A.B.C`, `A.B.*`, or `A.B.{C,D}`) (spec 13.2.1)
+        1. `A` is looked up in global scope
+        2. `B.C` (and `B.D`) or `B.*` is looked up. `A.B` must be a package.
+    """
+    # TODO: Default to copy False and check_builtin True with new instantiation/flattening
+
+    left_name, rest_of_name = _parse_str_or_ref(name)
+
+    # Simple Name Lookup first
+    found = _find_simple_name(
+        left_name,
+        scope,
+        search_imports=search_imports,
+        search_parent=search_parent,
+        check_builtin_classes=check_builtin_classes,
+        current_extends=current_extends,
+    )
+
+    # Composite Name Lookup if necessary
+    if found is not None and rest_of_name:
+        found = _find_rest_of_name(found, rest_of_name)
```

Comment:
Good question. I started with that, but since composite name lookup is essentially simple name lookup + rest of name lookup, this seemed to me like a more descriptive name. For example, to look up `A.B.C`, `_find_simple_name()` looks up `A` and `_find_rest_of_name()` looks up `B.C`. So `find_name()` is really the composite name lookup entry point. I think the main issue here is probably the comment more than the code.

Reply:
I updated the comment to hopefully make it clearer. Will show up in the next force push, but waiting on that until we have resolved all the comments from this review.